src/test/java/org/opensearch/sdk/sample/helloworld/TestHelloWorldExtension.java

```diff
@@ -49,4 +51,17 @@ public void testExtensionRestHandlers() {
         assertEquals(4, routes.size());
     }
 
+    @Test
+    public void testImplimentedInterfacesByExtension() {
+        Class<?>[] interfaces = extension.getClass().getInterfaces();
+        List<String> interfacesOfOpenSearch = new ArrayList<String>();
+        for (Class<?> anInterface : interfaces) {
+            if (anInterface.getPackageName() == "org.opensearch.sdk") {
+                interfacesOfOpenSearch.add(anInterface.getSimpleName());
+            }
+        }
+
+        assertEquals(interfacesOfOpenSearch, new ArrayList<String>(Arrays.asList("EngineExtension")));
```

Comment:
I'm a little confused here, as `HelloWorldExtension` extends `BaseExtension` which implements `Extension` and `ActionExtension`, so we shouldn't see `EngineExtension` in the list here and this test case should fail.

More importantly, this test case is just repeating the internal code for another method, which doesn't really test that method.  I'd expect to see `List<String> interfacesOfOpenSearch = someMethodThatReturnsTheList()`.

Since that code is in the handler which has a field for the `ExtensionsRunner`, which itself has a field for the `Extension`, you may want to make the `ExtensionsRunner` do that code logic and return the list in a method which you test here.  I think `getExtensionImplementedInterfaces()` is a good name.

Reply:
Yes that makes more sense. I have added that logic to ExtensionRunner and have added test method in TestExtensionRunner.java but I have 2 doubts here.

1)  I see there that they are using ExtensionRunnerForTest class as extensionRunner( which extends extensionRunner class ) and while constructing it, we construct its super class by providing instance of BaseExtension as param. How can we instantiate an abstract class ? ( BaseExtension is declared as an abstract class )

2) when I debug the method testGetExtensionImplementedInterfaces, Ideally getExtension() method of ExtensionRunner should getinvoked as its not been overridden in ExtensionRunnerForTestClass. so it should give me the instance of BaseExtension which impliments ActionExtesion, But I am getting ExtensionRunnerForTest$1@7829 as an object and getting an empty list ( implementedInterfaces ) as output.

so as of now as coverage wise we are good. testGetExtensionImplementedInterfaces teste the method and I am asserting whether returned list is empty, so the test is passing.

I might lack some knowledge here, So whenever you get time after you comeback from vacation, help me resolve these doubts. 


Thanks

Reply:
For 1.  You can instantiate any interface or abstract class like `new FooInterface() { @Override implemented methods here };`.  Although I think between the `ExtensionRunnerForTest` which internally has the extension extending `BaseExtension` or `HelloWorldExtension` (see `TestHelloWorldExtension` class) you should be able to test that method.

For 2. Should be a `getExtension()` method on the extensions runner that returns the extension?

Reply:
When I originally answered I hadn't seen your code, but now that I have I see what's going on.  The `ExtensionsRunnerForTest` class is creating an anonymous subclass of `BaseExtension`. That's what you see.

And on further reading the docs, the `getInterfaces()` only gets the interfaces of that subclass, which aren't any; it doesn't `implement` anything, it `extends BaseExtension`.  So what we really need to do is use its `superClass()` method to retrieve `BaseExtension` which should have the appropriate extensions.

This needs to be done recursively of course, for multiple levels of inheritance.  So your code should:
 - create a collection (Set, probably) of classes
 - extract the `Class` of the `extension`
 - call `getInterfaces()` and add all (possibly empty) to the set
 - call `getSuperClass()`. 
   - If you get a `null` return, you've hit the `Object` class, and you're done
   - otherwise, return to the `getInterfaces()` step, add to the set, and check the next superclass

Writing code from memory without an IDE, entirely untested, but hopefully explaining the above logic... it might work, or not, and you might need to add generics like `Class<?>` for the collection API, feel free to write your own version!
```
Set<Class> interfaceSet = new HashSet<>();
Class clazz = extension.getClass();
do {
  interfaceSet.addAll(clazz.getInterfaces());
  clazz = clazz.getSuperclass();
} while (clazz != null);
```

Reply:
Hi @dbwiddis ,

Understood now..  yes the approach which you told perfectly covers the case. amazing that so many corner cases will be there in a small feature. thanks for your input. I will change this appropriately. Currently I am drowned in resolving merge conflicts. I should start a habit of regularly syncing my repo with central and pulling it to local.

Thanks and regards
Varun.

Reply:
Hi @dbwiddis ,

I have added the change but rebasing taking a lot of time.. there are around 8 to 9 commits to this branch when it was created and I had not merged the latest now getting lot of merge conflicts while rebasing and having little confusion which changes to take and which not to take.. I will update you once its done. In the extreme case If I am not able to rebase successfully , I have the diff of my change, Is it okay If I create fresh PR with latest ?

Thanks and Regards
Varun

Reply:
Totally fine to start with a fresh PR!